### ISSUE TITLE
Pipeline/Component Serialization

### ISSUE THREAD

**Issue by jeremyliweishih** [alteryx/evalml#161]
Estimated requirements

* Save models to file, and load models from file
* Saving/loading a model should produce an object which is 100% identical to the original, including model state and random seed
* Security: when we transmit serialized models over the network, it should be prohibitively difficult for a third-party with network privileges to inject malicious code into the serialized model

**Comment by dsherry** [comment]
We had to implement a workaround for the ensembling work in #1128 : the parameters dict for ensembles will contain pipeline instances (containing the parameters for each). Once pipelines as DAGs (#273) is complete, we'd like to update ensemble pipelines to have a graph where each sub-pipeline in the ensemble is an input to the final ensemble estimator. #1281 tracks doing that update.

So, heads up that if we start on serialization *before* we finish #273 and update the ensemble pipelines in #1281, this could make pipeline serialization harder, because we'd have to handle the case where there are pipeline instances in the parameters dict for the ensemble pipeline. Once we've addressed #273 and #1281, this complication will go away.

@angela97lin @eccabay FYI

**Comment by dsherry** [comment]
#1956 made it so that we can save pipelines using the `pickle` format. We also currently support save pipelines using cloudpickle.

We have another issue out to track saving/loading automl search via pickle. #2174 . We currently support that via cloudpickle.

Therefore, closing this issue!
